File: src/main/java/com/neutralplasma/virtusbot/commands/admin/SayCommand.java
package com.neutralplasma.virtusbot.commands.admin;

import com.jagrosh.jdautilities.command.CommandEvent;
import com.neutralplasma.virtusbot.commands.AdminCommand;
import com.neutralplasma.virtusbot.utils.AbstractChatUtil;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.Member;

import java.awt.*;
import java.util.HashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;

public class SayCommand extends AdminCommand {

    private HashMap<String, EmbedBuilder> savedMessages = new HashMap<>();

    public SayCommand(){
        this.name = "say";
        this.help = "Bot says stuff instead of you in embed.";
        this.arguments = "<SUBCOMMAND>";
    }

    @Override
    protected void execute(CommandEvent commandEvent) {
        String args = commandEvent.getArgs();
        String[] cargs = args.split(" ");
        Member member = commandEvent.getMember();
        if (cargs.length > 0){
            EmbedBuilder message = savedMessages.get(member.getId());
            if(message == null){
                message = new EmbedBuilder();
            }
            EmbedBuilder finalMessage = message;

            if(cargs[0].equalsIgnoreCase("settitle")){
                commandEvent.reply("Please send the title: ");

                AbstractChatUtil abstractChatUtil = new AbstractChatUtil(commandEvent.getAuthor(), chatInfo -> {
                    finalMessage.setTitle(chatInfo.getMessage());
                    savedMessages.put(member.getId(), finalMessage);
                }, commandEvent.getJDA());
                abstractChatUtil.setOnClose(() -> {
                    commandEvent.reply("DONE");
                });

            }else if (cargs[0].equalsIgnoreCase("addfield")) {
                commandEvent.reply("Please send the field title: ");
                AtomicReference<String> title = new AtomicReference<>("");
                AbstractChatUtil abstractChatUtil = new AbstractChatUtil(commandEvent.getAuthor(), chatInfo -> {
                    title.set(chatInfo.getMessage());

                }, commandEvent.getJDA());
                abstractChatUtil.setOnClose(() -> {
                    commandEvent.reply("Please provide the field content.");
                    AbstractChatUtil abstractChatUtil2 = new AbstractChatUtil(commandEvent.getAuthor(), chatInfo2 -> {
                        finalMessage.addField(title.get(), chatInfo2.getMessage().replace("\\n", System.lineSeparator()), false);
                        savedMessages.put(member.getId(), finalMessage);
                    }, commandEvent.getJDA());

                    abstractChatUtil2.setOnClose(() -> {
                        commandEvent.reply("DONE");
                    });
                });
            }else if (cargs[0].equalsIgnoreCase("color")){
                commandEvent.reply("Please send color in next format: RRR:GGG:BBB (255:255:255):");
                AbstractChatUtil abstractChatUtil = new AbstractChatUtil(commandEvent.getAuthor(), chatInfo -> {
                    String[] color = chatInfo.getMessage().split(":");
                    if(color.length > 2){
                        Color color1 = new Color(Integer.parseInt(color[0]),Integer.parseInt(color[1]),Integer.parseInt(color[2]));
                        finalMessage.setColor(color1);
                        savedMessages.put(member.getId(), finalMessage);

                    }else{
                        commandEvent.reply("Please follow the format: RRR:GGG:BBB");
                    }
                }, commandEvent.getJDA());
                abstractChatUtil.setOnClose(() -> {
                    commandEvent.reply("Finished.");
                });

            }else if (cargs[0].equalsIgnoreCase("clear")){
                savedMessages.remove(member.getId());
                commandEvent.reply("DONE");

            }else if (cargs[0].equalsIgnoreCase("send")){
                commandEvent.reply(finalMessage.build());

            }else{
                commandEvent.reply(getInfo().build());
            }
        }else{
            commandEvent.reply(getInfo().build());
        }
        //commandEvent.getMessage().delete().queueAfter(1L, TimeUnit.SECONDS);
    }

    public EmbedBuilder getInfo(){
        EmbedBuilder eb = new EmbedBuilder();
        eb.setTitle("Info:");
        eb.setColor(Color.magenta.brighter());
        eb.addField("Available subcommands",
                "**send** - Sends message\n" +
                "**color** - Set the color of the message border\n" +
                "**addfield** - Add the field to message\n" +
                "**removefield <id>** - Remove field from message\n" +
                "**settitle** - Set the title of message\n" +
                "**clear** - Clears the current saved message\n " +
                "**getfields** - Lists all the fields with their titles\n" +
                "**gettitle** - Get the current title\n " +
                "**preview** - Sends the preview message", false);
        return eb;
    }
}

File: src/main/java/com/neutralplasma/virtusbot/commands/admin/SayCommand.kt
package com.neutralplasma.virtusbot.commands.admin

import com.jagrosh.jdautilities.command.CommandEvent
import com.neutralplasma.virtusbot.commands.AdminCommand
import com.neutralplasma.virtusbot.utils.AbstractChatUtil
import net.dv8tion.jda.api.EmbedBuilder
import java.awt.Color
import java.util.*
import java.util.concurrent.atomic.AtomicReference

class SayCommand : AdminCommand() {
    private val savedMessages = HashMap<String, EmbedBuilder>()
    override fun execute(commandEvent: CommandEvent) {
        val args = commandEvent.args
        val cargs = args.split(" ".toRegex()).toTypedArray()
        val member = commandEvent.member
        if (cargs.size > 0) {
            var message = savedMessages[member.id]
            if (message == null) {
                message = EmbedBuilder()
            }
            val finalMessage: EmbedBuilder = message
            if (cargs[0].equals("settitle", ignoreCase = true)) {
                commandEvent.reply("Please send the title: ")
                val abstractChatUtil = AbstractChatUtil(commandEvent.author, { chatInfo: AbstractChatUtil.ChatConfirmEvent ->
                    finalMessage.setTitle(chatInfo.message)
                    savedMessages.put(member.id, finalMessage)
                }, commandEvent.jda)


                abstractChatUtil.onClose = {
                    commandEvent.reply("DONE")
                }

            } else if (cargs[0].equals("addfield", ignoreCase = true)) {
                commandEvent.reply("Please send the field title: ")
                val title = AtomicReference("")
                val abstractChatUtil = AbstractChatUtil(commandEvent.author, {
                    chatInfo: AbstractChatUtil.ChatConfirmEvent -> title.set(chatInfo.message)
                }, commandEvent.jda)

                abstractChatUtil.onClose = {
                    commandEvent.reply("Please provide the field content.")
                    val abstractChatUtil2 = AbstractChatUtil(commandEvent.author, { chatInfo2: AbstractChatUtil.ChatConfirmEvent ->
                        finalMessage.addField(title.get(), chatInfo2.message.replace("\\n", System.lineSeparator()), false)
                        savedMessages.put(member.id, finalMessage)
                    }, commandEvent.jda)

                    abstractChatUtil2.onClose = { commandEvent.reply("DONE") }

                }
            } else if (cargs[0].equals("color", ignoreCase = true)) {
                commandEvent.reply("Please send color in next format: RRR:GGG:BBB (255:255:255):")
                val abstractChatUtil = AbstractChatUtil(commandEvent.author, { chatInfo: AbstractChatUtil.ChatConfirmEvent ->
                    val color = chatInfo.message.split(":".toRegex()).toTypedArray()
                    if (color.size > 2) {
                        val color1 = Color(color[0].toInt(), color[1].toInt(), color[2].toInt())
                        finalMessage.setColor(color1)
                        savedMessages.put(member.id, finalMessage)
                    } else {
                        commandEvent.reply("Please follow the format: RRR:GGG:BBB")
                    }
                }, commandEvent.jda)

                abstractChatUtil.onClose =  { commandEvent.reply("Finished.") }

            } else if (cargs[0].equals("clear", ignoreCase = true)) {
                savedMessages.remove(member.id)
                commandEvent.reply("DONE")
            } else if (cargs[0].equals("send", ignoreCase = true)) {
                commandEvent.reply(finalMessage.build())
            } else {
                commandEvent.reply(info.build())
            }
        } else {
            commandEvent.reply(info.build())
        }
        //commandEvent.getMessage().delete().queueAfter(1L, TimeUnit.SECONDS);
    }

    val info: EmbedBuilder
        get() {
            val eb = EmbedBuilder()
            eb.setTitle("Info:")
            eb.setColor(Color.magenta.brighter())
            eb.addField("Available subcommands",
                    """**send** - Sends message
**color** - Set the color of the message border
**addfield** - Add the field to message
**removefield <id>** - Remove field from message
**settitle** - Set the title of message
**clear** - Clears the current saved message
 **getfields** - Lists all the fields with their titles
**gettitle** - Get the current title
 **preview** - Sends the preview message""", false)
            return eb
        }

    init {
        name = "say"
        help = "Bot says stuff instead of you in embed."
        arguments = "<SUBCOMMAND>"
    }
}